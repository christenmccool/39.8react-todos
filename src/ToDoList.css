.ToDoList {
  text-align: center;
}

.ToDoList-title {
  margin: 20px auto 20px auto;
}

.ToDoList-newform {
  margin-bottom: 40px;
} 