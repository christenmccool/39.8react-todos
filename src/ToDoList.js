import React, {useEffect, useState} from 'react';
import { v4 as uuid } from 'uuid';
import './ToDoList.css';
import ToDo from './ToDo';
import NewToDoForm from './NewToDoForm';


const ToDoList = () => {
  const initalState = JSON.parse(localStorage.getItem("toDoList")) || [];
  const [toDoList, setToDoList] = useState(initalState);
  useEffect(() => localStorage.setItem("toDoList", JSON.stringify(toDoList)), [toDoList]);

  const renderToDos = () => (
    toDoList.map(toDo => (
      <ToDo 
        id={toDo.id}
        key={toDo.id}
        task={toDo.task} 
        priority={toDo.priority} 
        complete={toDo.complete}
        fadeout={toDo.fadeout}
        toEdit={toDo.toEdit}
        deleteToDo={() => deleteToDo(toDo.id)}
        toggleProp={(prop) => toggleProp(toDo.id, prop)}
        editToDo={editToDo}
      />
    ))
  );

  const deleteToDo = (id) => {
    setToDoList(toDoList.filter(toDo => toDo.id !== id));
  }

  const addToDo = (taskObj) => {
    setToDoList([...toDoList, {...taskObj, complete: false, fadeout: false, toEdit: false, id: uuid()}]);

    // localStorage.setItem("toDoList", JSON.stringify(toDoList));
  }

  const toggleProp = (id, prop) => {
    const targetToDo = toDoList.find(toDo => toDo.id === id);

    targetToDo[prop] = !targetToDo[prop];
    setToDoList(toDoList.map(todo => todo.id === id ? targetToDo : todo));
  }

  const editToDo = (id, newData) => {
    const targetToDo = toDoList.find(toDo => toDo.id === id);
    const editedToDo = {...targetToDo, ...newData, toEdit: false};

    setToDoList(toDoList.map(todo => todo.id === id ? editedToDo : todo));
  }


  return (
    <div className="ToDoList">
      <h1 className="ToDoList-title">To Do List</h1>

      <div className="ToDoList-newform">
        <NewToDoForm 
          addToDo={addToDo}
        />
      </div>

      <div className="ToDoList-list">
        {renderToDos()}
      </div>
    </div>
  )
}

export default ToDoList;