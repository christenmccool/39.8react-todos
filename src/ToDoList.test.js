import { render, screen, fireEvent } from '@testing-library/react';
import ToDoList from './ToDoList';

test('if renders without crashing', () => {
  render(<ToDoList />);
});

test("if matches snapshot", () => {
  const {asFragment} = render(<ToDoList />);
  expect(asFragment()).toMatchSnapshot();
})

test("if new todo is added", () => {
  const { queryByLabelText, queryByText, queryByRole, debug } = render(<ToDoList />);
  
  expect(queryByText("Sweep floor")).not.toBeInTheDocument();
  expect(queryByText("Complete")).not.toBeInTheDocument();
  expect(queryByText("X")).not.toBeInTheDocument();
  expect(queryByText("Edit")).not.toBeInTheDocument();

  const task = queryByLabelText("New task");
  const low = queryByRole("radio", {name: "Low"});
  const medium = queryByRole("radio", {name: "Medium"});
  const high = queryByRole("radio", {name: "High"});
  const button = queryByText("Add");

  fireEvent.change(task, { target: { value: 'Sweep floor' } });
  fireEvent.change(low, { target: { checked: true } });
  fireEvent.click(button);

  expect(queryByText("Sweep floor")).toBeInTheDocument();
  expect(queryByText("Complete")).toBeInTheDocument();
  expect(queryByText("X")).toBeInTheDocument();
  expect(queryByText("Edit")).toBeInTheDocument();
})

test("if todo is deleted", () => {
  const { queryByLabelText, queryByText, queryByRole, debug } = render(<ToDoList />);
  
  const task = queryByLabelText("New task");
  const low = queryByRole("radio", {name: "Low"});
  const medium = queryByRole("radio", {name: "Medium"});
  const high = queryByRole("radio", {name: "High"});
  const button = queryByText("Add");

  fireEvent.change(task, { target: { value: 'Sweep floor' } });
  fireEvent.change(low, { target: { checked: true } });
  fireEvent.click(button);

  expect(queryByText("Sweep floor")).toBeInTheDocument();
  expect(queryByText("Complete")).toBeInTheDocument();
  expect(queryByText("X")).toBeInTheDocument();
  expect(queryByText("Edit")).toBeInTheDocument();

  const deleteBtn = queryByText("X");
  fireEvent.click(deleteBtn);
  
  setTimeout(() => {
    expect(queryByText("Sweep floor")).not.toBeInTheDocument();
    expect(queryByText("Complete")).not.toBeInTheDocument();
    expect(queryByText("X")).not.toBeInTheDocument();
    expect(queryByText("Edit")).not.toBeInTheDocument();
  }, 3000);

})