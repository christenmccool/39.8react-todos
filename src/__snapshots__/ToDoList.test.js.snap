// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`if matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="ToDoList"
  >
    <h1
      class="ToDoList-title"
    >
      To Do List
    </h1>
    <div
      class="ToDoList-newform"
    >
      <form
        class="NewToDoForm"
      >
        <label
          class="NewToDoForm-label"
          for="task"
        >
          New task
        </label>
        <input
          id="task"
          name="task"
          type="text"
          value=""
        />
        <fieldset
          class="NewToDoForm-priority"
        >
          <legend>
            Priority
          </legend>
          <input
            id="choice3"
            name="priority"
            type="radio"
            value="low"
          />
          <label
            class="NewToDoForm-label"
            for="choice3"
          >
            Low
          </label>
          <input
            checked=""
            id="choice2"
            name="priority"
            type="radio"
            value="medium"
          />
          <label
            class="NewToDoForm-label"
            for="choice2"
          >
            Medium
          </label>
          <input
            id="choice1"
            name="priority"
            type="radio"
            value="high"
          />
          <label
            class="NewToDoForm-label"
            for="choice1"
          >
            High
          </label>
        </fieldset>
        <button
          class="NewToDoForm-btn"
        >
          Add
        </button>
      </form>
    </div>
    <div
      class="ToDoList-list"
    />
  </div>
</DocumentFragment>
`;
